[buienradar] Buienradar Binding initial contribution

gedejong opened the pull request:
Hi, this is the initial commit for the buienradar binding. Buienradar is a Dutch/Belgian website which uses radar information to get precipitation information and predictions. This binding simply polls the API every 5 minutes to retrieve current and predicted situations. This can be used to control sprinklers or advice on windows which are inadvertently open during a storm.
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific addon: Mention the addon shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the addon README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/development/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/development/bindings.html#static-code-analysis
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing/contributing.html#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It's a good practice to add an URL to your built JAR in this Pull Request description,
so it's easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it's reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

gedejong commented:
What do I need to do to get this PR approved?

wborn commented:
Thanks for the binding @gedejong! We're updating the current bindings so they build with bnd. Can you also migrate this binding to a bnd build (#5005)? See for instance the updates of the AVM Fritz Binding in #5050.

J-N-K commented:
Basically you need to copy the src-folder, NOTICE and resend from addons/bindings/org.openhab... to bundles/org.openhab..., copy pom.xml, .project, .classpath and the .settings-folder from the astro or dmx binding to your new bindings folder, change pom.xml and .project with your new bindings name.

Then adapt the feature.xml, bom, CODEOWNERS, addons/binding/pom.xml (remove your binding), bundles/pom.xml (add your binding). For feature,bom,CODEOWNERS take astro or DMX as example.

If you have any question, feel free to ask.

J-N-K commented on addons/binding/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/HueBridge.java:
```diff
@@ -12,7 +12,7 @@
 package org.openhab.binding.hue.internal;
 
-import java.io.IOException;
+ import java.io.IOException;
```
Seems to be unrelated, please remove

gedejong replied:
Fixed

J-N-K commented on addons/binding/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    private Runnable getRunnable() {
+        return () -> {
```
You could also make this a `private void refresh()`, dropping the lambda-part and use `this::refresh` in `scheduleWithFixedDelay`.

gedejong replied:
Fixed

J-N-K commented on addons/binding/org.openhab.binding.buienradar/release.properties:
```diff
@@ -0,0 +1,11 @@
+#release configuration
```
Please remove this file

gedejong replied:
Fixed

bjoernbrings commented:
Sorry for the silly question, but before reviewing: How about the licensing of the parsed data? Though I don't speak netherlands the disclaimer on the homepage does (with google translate) not read like this ...

J-N-K commented:
I'm not a lawyer (and don't speak dutch) but where do you see a problem here? As far as I can see the use of the data is forbidden (without permission) for commercial use. OH is not using the data at all, it merely translates this data. So clearly the use of this binding in a Commercial Environment would be against ther Terms of Use, but for private Households it shpuld be ok.

bjoernbrings commented on addons/binding/org.openhab.binding.buienradar/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,223 @@
+		</channels>
+
+		<config-description>
+			<parameter name="longitude" type="decimal" required="true">
```
Wouldn't it be more intuitive to have first lat, then lon (EPSG:4326)?

gedejong commented on addons/binding/org.openhab.binding.buienradar/README.md:
```diff
@@ -0,0 +1,14 @@
+
+## Discovery
+
+No auto-discovery is possible.
```
True, I'll add that later. First wanted to see if the code was ok and this fixes an actual problem.

gedejong commented on addons/binding/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java:
```diff
@@ -0,0 +1,120 @@
+        updateStatus(ThingStatus.UNKNOWN);
+
+        if (listenableFuture == null || listenableFuture.isCancelled()) {
+            listenableFuture = scheduler.scheduleWithFixedDelay(getRunnable(), 0L, 5L, MINUTES);
```
Fixed

gedejong commented on addons/binding/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java:
```diff
@@ -0,0 +1,140 @@
+            // Intensity in mm / hour = 10^((value-109)/32)
+            return Math.pow(10.0, (Integer.parseInt(trimmedStr) - 109) / 32.0);
+        } catch (NumberFormatException e) {
+            throw new BuienradarParseException("Could not parse intensityStr", e);
```
Not at all, it was part of the first version (before refactor into methods).

gedejong commented on addons/binding/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java:
```diff
@@ -0,0 +1,140 @@
+    public static double parseIntensity(String intensityStr) throws BuienradarParseException {
+        assert (intensityStr.length() == 3);
+        final int lastZero = intensityStr.lastIndexOf('0');
+        final String trimmedStr = lastZero == -1 || lastZero == 2 ? intensityStr : intensityStr.substring(lastZero + 1);
```
This was indeed a bug... It was not necessary. I thought 010 might be parsed as octal. Fixed.

gedejong commented on addons/binding/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java:
```diff
@@ -0,0 +1,140 @@
+        final int minute = Integer.parseInt(timeElements[1]);
+        final LocalTime time = LocalTime.of(hour, minute);
+
+        final ZonedDateTime tryDateTime = time.atDate(now.toLocalDate()).atZone(ZoneId.of("Europe/Amsterdam"));
```
The time that buienradar returns is always at timezone `Europe/Amsterdam` afaik. I did do some extra checks wrt the date, so we always calculate in the correct timezone.

gedejong commented on addons/binding/org.openhab.binding.buienradar/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,12 @@
+
+	<name>Buienradar Binding</name>
+	<description>
+	The Buienradar Binding periodically (5 minutes) retrieves results from the Buienradar API. For personal use only.
```
License restrictions are copied from the original API in the README.md. 5 min has become a setting.

gedejong commented on addons/binding/pom.xml:
```diff
@@ -1,4 +1,5 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?><project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
Indeed, now obsolete. Not sure what changed this file though.

gedejong commented on addons/binding/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java:
```diff
@@ -0,0 +1,140 @@
+        final List<Prediction> predictions = new ArrayList<Prediction>(24);
+        final List<String> errors = new LinkedList<String>();
+        logger.debug("Returned result from buienradar: {}", result);
+        for (String line : result.split("\n")) {
```
Fixed.

gedejong commented on addons/binding/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java:
```diff
@@ -0,0 +1,140 @@
+        final String trimmedStr = lastZero == -1 || lastZero == 2 ? intensityStr : intensityStr.substring(lastZero + 1);
+        try {
+            // Intensity in mm / hour = 10^((value-109)/32)
+            return Math.pow(10.0, (Integer.parseInt(trimmedStr) - 109) / 32.0);
```
Migrated to BigDecimal with two significant digits. This would be rounded to 0.00 mm/h.

gedejong replied:
Looking at the actual results in my Grafana dashboard, I wonder if 1 significant digit would make more sense. Let me do a bit more research in influxdb.

gedejong replied:
Hmmm, this has some 'strange' values at the top, with 0.0005 mm/hr. Could it be that significant, or can we safely ignore it?

`influx -username 'bigsmall' -password *** -database 'bigsmall' -execute 'select * from BuienradarForecast_CurrentRainfall' | tail -n +4 | cut -d' ' -f2- | sort | uniq -c`
`

```
     1 0
    106 0.0003924189758484536
     42 0.0005232991146814947
     30 0.0006042963902381329
     25 0.0007498942093324559
     99 0.1
     63 0.2053525026457146
     52 0.29427271762092816
     51 0.4869675251658631
     23 1.2409377607517196
     23 1.333521432163324
     22 1.5399265260594919
     11 1.6548170999431815
     13 1.7782794100389228
     40 2.0535250264571463
      2 2.5482967479793466
      1 2.7384196342643614
      5 3.3982083289425593
      4 3.651741272548377
      2 3.9241897584845358
      1 4.531583637600818
      6 4.869675251658631
      3 5.232991146814947
      1 9.30572040929699
```

gedejong commented:
Ok, binding is migrated to the new format and I worked on the CRs. Only needs a more extensive README, but I never configure my bindings via text-files (anymore). Anyone care to add it who is testing this binding as well?

gedejong commented:
@J-N-K perhaps this is useful for you: https://gist.github.com/gedejong/bcdec704acf8ff2bf804954ebe23ba85

davidgraeff commented on bundles/org.openhab.binding.buienradar/.settings/org.eclipse.core.resources.prefs:
```diff
@@ -0,0 +1,4 @@
+eclipse.preferences.version=1
```
Can you please remove the .settings directory? We have put that on the git ignore list recently as well.

gedejong replied:
Done

davidgraeff commented on bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java:
```diff
@@ -0,0 +1,120 @@
+    public static final Unit<Speed> MM_PER_HOUR = Units.METRE.divide(1000).divide(Units.HOUR).asType(Speed.class);
+
+    @Nullable
+    private BuienradarConfiguration config;
```
Please use the pattern `private @Nullable ..` for the nullable annotation

gedejong replied:
Done

Hilbrand commented on bundles/org.openhab.binding.buienradar/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,230 @@
+
+		<config-description>
+			
+			<parameter name="latitude" type="decimal" required="true">
```
There is standard support for this in openHAB. Use `<context>location</location>` with a single parameter of type String. See `openweathermap` binding for how this works. It makes it possible to select the location on a map in the ui.

gedejong replied:
Fixed.

gedejong commented on bundles/org.openhab.binding.buienradar/.settings/org.eclipse.m2e.core.prefs:
```diff
@@ -0,0 +1,4 @@
+activeProfiles=
+eclipse.preferences.version=1
```
Removed the `.settings` directory.

gedejong commented on bundles/org.openhab.binding.buienradar/README.md:
```diff
@@ -0,0 +1,14 @@
+
+## Supported Things
+
+The webservice supports one thing, which can be added manually via the web interface. The thing needs longitude and latitude
```
Fixed

gedejong commented on bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java:
```diff
@@ -0,0 +1,120 @@
+
+    public static final Unit<Speed> MM_PER_HOUR = Units.METRE.divide(1000).divide(Units.HOUR).asType(Speed.class);
+
+    @Nullable
```
Fixed

gedejong commented on bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java:
```diff
@@ -0,0 +1,120 @@
+    private final PredictionAPI client = new BuienradarPredictionAPI();
+
+    @Nullable
+    private ScheduledFuture<?> listenableFuture;
```
Fixed

gedejong commented on bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java:
```diff
@@ -0,0 +1,120 @@
+
+            updateStatus(ThingStatus.ONLINE);
+        } catch (IOException e) {
+            logger.error("Cannot retrieve predictions", e);
```
Fixed.

gedejong commented on bundles/org.openhab.binding.buienradar/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,230 @@
+			</parameter>
+			
+			<parameter name="longitude" type="decimal" required="true">
+                <label>Longitude</label>
```
Fixed.

gedejong commented on bundles/pom.xml:
```diff
@@ -1,22 +1,16 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
I used an XML processor to change the file automatically.

gedejong replied:
Fixed

gedejong commented on bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java:
```diff
@@ -0,0 +1,120 @@
+
+    private final Logger logger = LoggerFactory.getLogger(BuienradarHandler.class);
+
+    public static final Unit<Speed> MM_PER_HOUR = Units.METRE.divide(1000).divide(Units.HOUR).asType(Speed.class);
```
Somehow I missed that. Got directed to SIUnits. The units system is quite opaque for a beginning Openhab developer (and this is from someone with extensive knowledge of a similar system in Scala). I'd expect to be able to *need* to write it this way (in terms of base units), since otherwise you get an explosion of different units.

gedejong replied:
It seems this was added in the master branch, but I'm running this binding on 2.5.0M1. This would mean I can't run my own binding on my Odroid machine.

davidgraeff replied:
No need to change then. It is absolutely fine to only use API of a released core version.

Hilbrand replied:
It would only mean you can't run the binding compiled with the sources in what will be snapshot version. Nobody forbids you to locally modify the source to use the unit locally, compile it and run it. And even if it's kept anybody can put in a pr with this changed to SmarthomeUnits and than it will be modified anyway. So I see no reason why it should not be modified in this pr.

gedejong replied:
Don't want to argue, but at one point you'd expect a stable base API, with a certain amount of backward and forward compatibility. I mean, no offense, but I spent a lot of my spare time writing this binding, searching again and again to do things 'the right way': setting up an IDE I don't know well (and frankly runs very slowly on my Mac), going through many CR comments, update my PR because I know I'm a noob in the ecosystem. Why? Because I want to do sth back to the ecosystem, having run OpenHAB for at least 4 years. You're friendly and all, don't get me wrong, and I understand you want to get things clean and of high quality, so would I want to have it as well.

However, the actual logic of this PR is around 30 lines. I wrote it in Scala as well, which is literally 7 lines including configuration. For you, writing such a binding would take 10 minutes, but for someone new to the ecosystem it takes 3-4 evenings, plus time for the CRs and the move to the new bundles format.

Now, I want to get this PR finished, test it on my own machine and not have to go back and forth with a  MILLIMETRE_PER_HOUR while I start implementing the auto-discovery feature.

I'd love to write my usual `100.millimeter / hour` (literal Scala syntax), but it took me at least an hour to figure out the proper syntax for `public static final Unit<Speed> MM_PER_HOUR = Units.METRE.divide(1000).divide(Units.HOUR).asType(Speed.class); `

So, it would be nice if you can accept the PR without this change.

davidgraeff replied:
openHAB allowed Groovy once and it took ages to get rid of it again. Fortunately java gets more and more relaxed with newer versions. In Java 11 you can use "var" instead of repeating the type once more.

Regarding approval. I'm sure that @Hilbrand will not block the merge because of this one line, he just wants to reduce technical debts that accumulate if we merge code that is already "outdated". At the same time contributors like you cannot be expected to develop against OH core master, you are on the safe site.

gedejong replied:
Ok, sorry, I'm reading my comment again and it reads like a rant.

(p.s. I'm not claiming to use Scala. I would write my own system from scratch in Scala if I wanted that. Still, I don't think one should compare Groovy with Scala, they are fundamentally different wrt type-safety.)

Hilbrand replied:
No I won't block :smile: I do feel your pain. When I started I was asked twice: 'o btw can you use this new api instead' only to find out it wasn't complete and I had to add pr's to core. So can you at least leave a comment as suggested and use the same naming for the variable?

Hilbrand commented on bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java:
```diff
@@ -0,0 +1,152 @@
+
+    private static final String BASE_ADDRESS = "https://gpsgadget.buienradar.nl/data/raintext";
+
+    private static final int TIMEOUT = 15000;
```
15 seconds is a very long timeout. Especially considering this runs on a limited scheduled pool. So if there is something wrong with the service and it times. So maybe 1, 2 or 3 seconds is a better option. Also prefer to use the unit in the constant name: `TIMEOUT_MILLISECONDS`

gedejong replied:
The buienradar service is sometimes very slow, that's why I increased it to 15000. I'll place it back to 3 seconds for now and see how it runs.

(btw, seems like a good incentive to move these kind of HTTP calls to their own HTTP client pool and return a Future instead. Haven't done professional programming in Java for a while, so I don't know how it compares with Scala though.)

davidgraeff replied:
The httpclient that we use is capable of async requests with timeouts (internally hopefully with java.nio implemented if not at least it is using its own thread pool). You might indeed want to use that instead

gedejong replied:
The common usage of the httpclient is wrapped many times in `HttpUtil` without creating an orthogonal abstraction (see `org.eclipse.smarthome.io.net.http.HttpUtil.executeUrlAndGetReponse(String, String, Properties, InputStream, String, int, String, Integer, String, String, String)`). So I'd have to implement all this logic again, but then in an asynchronous manner. 

So, in a certain way I cannot use the underlying HTTP client without reimplementing that logic, since I'd break proxy support, expected debug logging, auth headers and shared HTTP client pools (although I think that's not applicable here, since we're sharing via Jetty).

Also, certain logic wrt proxying is hidden (`org.eclipse.smarthome.io.net.http.HttpUtil.prepareProxyParams()`), so even if I would copy it,  I might break forward compatibility, since this method could be changed in the future.

davidgraeff replied:
Oh, I don't mean to use `org.eclipse.smarthome.io.net.http.HttpUtil`. I have never done that and don't understand its use case to be honest. There is a `HttpClientFactory`. Get that as a service in your thing factory (or discovery services) and pass it to where you need it. Let the factory create a httpclient for you and voila

gedejong replied:
Aha, well, I'll migrate to that in the next PR with auto discovery. Out of time for this week... I picked up the `HttpUtil` from elsewhere (forgot which binding), thinking that this is 'normal' path. Seeing the proxy code using properties, I assumed other avenues are not wanted. There are *so* *many* *HTTP* client libraries out there... Simply don't know which one to use or how to even find information on which one is preferable.

gedejong commented on bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java:
```diff
@@ -0,0 +1,152 @@
+            }
+        }
+        if (!errors.isEmpty()) {
+            logger.error("Could not parse all results: " + errors.stream().collect(Collectors.joining(", ")));
```
Fixed and searched.

gedejong commented on bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java:
```diff
@@ -0,0 +1,152 @@
+
+    @Override
+    public List<Prediction> getPredictions(double lat, double lon) throws IOException {
+        final String address = String.format(BASE_ADDRESS + "?lat=%.6f&lon=%.6f", lat, lon);
```
Fixed, good catch. Normally I would not format urls like this, but use a URI/URL interpolation library instead (so it is locale independent and able to escape things correctly). However, I couldn't find such a utility.

gedejong commented on bundles/org.openhab.binding.buienradar/src/main/resources/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,12 @@
+
+	<name>Buienradar Binding</name>
+	<description>
+	The Buienradar Binding periodically (5 minutes) retrieves results from the Buienradar API. For personal use only.
```
Done

gedejong commented:
Alright, I think I've addressed all the comments. Please check the current results and see if this is mergable! Thanks for all the CR, much appreciated.

davidgraeff commented:
Travis doesn't agree: `(osgi.wiring.package=org.openhab.core.library.types)`
Shouldn't that be `org.eclipse.smarthome.core.library.types`?

Hilbrand commented on features/karaf/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -103,6 +103,11 @@
         <bundle start-level="80">mvn:org.openhab.addons.bundles/org.openhab.binding.bosesoundtouch/${project.version}</bundle>
     </feature>
 
+    <feature name="openhab-binding-buienadar" description="Buienradar Binding" version="${project.version}">
```
Typo in name.
```suggestion
    <feature name="openhab-binding-buienradar" description="Buienradar Binding" version="${project.version}">
```

J-N-K commented on bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java:
```diff
@@ -0,0 +1,156 @@
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.smarthome.io.net.http.HttpUtil;
+import org.openhab.core.library.types.PointType;
```
```suggestion
import org.eclipse.smarthome.core.library.types.PointType;
```

This should fix the build error

J-N-K commented:
@gedejong, could you fix the conflicts?

Hilbrand commented:
@openhab/2-x-add-ons-maintainers Can we merge the new binding buienradar?

J-N-K commented:
Manually verified that all `Signed-off-by`s are present

J-N-K commented:
Jenkins fails unrelated.
